tests/src/test/scala/pureconfig/ConfigSourceSuite.scala

```diff
@@ -0,0 +1,199 @@
+package pureconfig
+
+import java.io.IOException
+
+import com.typesafe.config.{ Config, ConfigFactory, ConfigValueType }
+import pureconfig.PathUtils._
+import pureconfig.error._
+import pureconfig.generic.auto._
+
+class ConfigSourceSuite extends BaseSuite {
+
+  behavior of "ConfigSource"
+
+  it should "load from application.conf and reference.conf by default" in {
+    case class Conf(d: Double, i: Int, s: String)
+    ConfigSource.default.load[Conf] shouldBe Right(Conf(0D, 0, "app_value"))
+  }
+
+  it should "allow reading at a namespace" in {
+    case class Conf(f: Float)
+    ConfigSource.default.at("foo").load[Conf] shouldBe Right(Conf(3.0F))
+  }
+
+  it should "allow reading from a config object" in {
+    case class Conf(d: Double, i: Int)
+    val conf = ConfigFactory.parseString("{ d: 0.5, i: 10 }")
+    ConfigSource.fromConfig(conf).load[Conf] shouldBe Right(Conf(0.5D, 10))
+  }
+
+  it should "allow reading from a config object at a namespace" in {
+    case class Conf(f: Float)
+    val conf = ConfigFactory.parseString("foo.bar { f: 1.0 }")
+    ConfigSource.fromConfig(conf).at("foo.bar").load[Conf] shouldBe Right(Conf(1.0F))
+    ConfigSource.fromConfig(conf).at("bar.foo").load[Conf] should failWith(KeyNotFound("bar", Set.empty), "")
+  }
+
+  it should "allow reading scalar values from a config object at a namespace" in {
+    val conf = ConfigFactory.parseString("foo { bar { f: 1.0 }, baz: 3.4 }")
+    val source = ConfigSource.fromConfig(conf)
+
+    source.at("foo.bar.f").load[Float] shouldBe Right(1.0F)
+    source.at("foo.bar.h").load[Float] should failWith(KeyNotFound("h", Set.empty), "foo.bar")
+    source.at("bar.foo.f").load[Float] should failWith(KeyNotFound("bar", Set.empty), "")
+    source.at("foo.baz.f").load[Float] should failWith(
+      WrongType(ConfigValueType.NUMBER, Set(ConfigValueType.OBJECT)), "foo.baz")
+
+    source.at("foo.bar.f").load[Option[Float]] shouldBe Right(Some(1.0F))
+    source.at("foo.bar.h").load[Option[Float]] should failWith(KeyNotFound("h", Set.empty), "foo.bar")
+    source.at("bar.foo.f").load[Option[Float]] should failWith(KeyNotFound("bar", Set.empty), "")
+    source.at("foo.baz.f").load[Option[Float]] should failWith(
+      WrongType(ConfigValueType.NUMBER, Set(ConfigValueType.OBJECT)), "foo.baz")
```

Comment:
Can you also add a similar test that uses multiple `.at` calls (e.g. `source.at("foo").at("bar").at("f")`) and checks that failures have the appropriate path?

Reply:
1b594a0